Actions/Github-Helper.psm1

```diff
@@ -708,20 +708,234 @@ function Set-JsonContentLF {
     }
 }
 
+<#
+    Checks if a CICD workflow run is successful.
+    CICD run is successful if all the build jobs are successful.
+#>
+function IsCICDSuccessful {
+    Param(
+        [Parameter(Mandatory = $true)]
+        [string] $token,
+        [Parameter(Mandatory = $true)]
+        [string] $api_url,
+        [Parameter(Mandatory = $true)]
+        [string] $repository,
+        [Parameter(Mandatory = $true)]
+        [string] $CICDRunId
+    )
+
+    $headers = GetHeader -token $token
+    $per_page = 100
+    $page = 1
+
+    $isSuccessful = $true
+
+    while($true) {
+        $jobsURI = "$api_url/repos/$repository/actions/runs/$CICDRunId/jobs?per_page=$per_page&page=$page"
+        Write-Host "- $jobsURI"
+        $workflowJobs = InvokeWebRequest -Headers $headers -Uri $runsURI | ConvertFrom-Json
+
+        if($workflowJobs.jobs.Count -eq 0) {
+            # No more jobs, breaking out of the loop
+            break
+        }
+        $buildJobs = @($workflowJobs.jobs | Where-Object { $_.name.StartsWith('Build ') })
+
+        if($buildJobs.conclusion -ne 'success') {
+            # If there is a build job that is not successful, the CICD run is not successful
+            $isSuccessful = $false
+        }
+
+        if(-not $isSuccessful) {
+            # CICD run is not successful, breaking out of the loop
+            break
+        }
+
+        $page += 1
+    }
+
+    return $isSuccessful
+}
+
+<#
+    Gets the last successful CICD run ID for the specified repository and branch.
+    Successful CICD runs are those that have a workflow run named ' CI/CD' and successfully built all the projects.
+
+    If no successful CICD run is found, 0 is returned.
+#>
+function FindLatestSuccessfulCICDRun {
+    Param(
+        [Parameter(Mandatory = $true)]
+        [string] $token,
+        [Parameter(Mandatory = $true)]
+        [string] $api_url,
+        [Parameter(Mandatory = $true)]
+        [string] $repository,
+        [Parameter(Mandatory = $true)]
+        [string] $branch
+    )
+
+    $headers = GetHeader -token $token
+    $lastSuccessfulCICDRun = 0
+    $per_page = 100
+    $page = 1
+
+    Write-Host "Finding latest successful CICD run for branch $branch in repository $repository"
+
+    # Get the latest CICD workflow run
+    while($true) {
+        $runsURI = "$api_url/repos/$repository/actions/runs?per_page=$per_page&page=$page&exclude_pull_requests=true&status=completed&branch=$branch"
+        Write-Host "- $runsURI"
+        $workflowRuns = InvokeWebRequest -Headers $headers -Uri $runsURI | ConvertFrom-Json
+
+        if($workflowRuns.workflow_runs.Count -eq 0) {
+            # No more workflow runs, breaking out of the loop
+            break
+        }
+
+        $CICDRuns = @($workflowRuns.workflow_runs | Where-Object { $_.name -eq ' CI/CD' })
+
+        foreach($CICDRun in $CICDRuns) {
```

Comment:
Could we check whether the workflow is marked successful before checking the individual jobs? (in that case, there is no need for checking whether the build jobs are successful)

Reply:
Oh, nice catch. I thought I had the check for successful workflow run.

As for the `in_progress` status, the query filters on `completed`, so in progress runs will be omitted.